modules/phase_field/src/postprocessors/FeatureFloodCount.C

```diff
@@ -467,129 +468,76 @@ FeatureFloodCount::mergeSets(bool use_periodic_boundary_info)
-              it2->_merged = true;
-
-              // Something was merged so we'll need to repeat this loop
-              region_merged = true;
-            }
-          } // it2 loop
-
-          // Don't increment if we had a merge, we need to retry earlier candidates again
-          if (!region_merged)
-            ++it1;
-          else
-            goto REPEAT_MERGE_LOOPS;
-
-        } // it1 loop
-      } // rank2 loop
-    } // rank1 loop
-  } // map loop
+             )
+           )
+        {
+          it2->merge(std::move(*it1));
 
+          // Insert the new entity at the end of the list so that it may be checked against all other partial features again
+          _partial_feature_sets[map_num].emplace_back(std::move(*it2));
 
-  for (auto map_num = decltype(_maps_size)(0); map_num < _maps_size; ++map_num)
-    _feature_sets[map_num].clear();
+          /**
+           * Now remove both halves the merged features: it2 contains the "moved" feature cell just inserted
+           * at the back of the list, it1 contains the mostly empty other half. We have to be careful about the
+           * order in which these two elements are deleted. We delete it2 first since we don't care where its
+           * iterator points after the deletion. We are going to break out of this loop anyway. If we delete
+           * it1 first, it may end up pointing at the same location as it2 which after the second deletion would
+           * cause both of the iterators to be invalidated.
+           */
+          _partial_feature_sets[map_num].erase(it2);
+          it1 = _partial_feature_sets[map_num].erase(it1); // it1 is incremented here!
 
-  // Now consolidate the data structure
-  _feature_count = 0;
-  for (auto rank = decltype(_n_procs)(0); rank < _n_procs; ++rank)
-    for (auto map_num = decltype(_maps_size)(0); map_num < _maps_size; ++map_num)
-    {
-      for (auto & feature : _partial_feature_sets[rank][map_num])
-      {
-        if (!feature._merged)
-        {
-          _feature_sets[map_num].emplace_back(std::move(feature));
-          ++_feature_count;
+          // A merge occurred, this is used to determine whether or not we increment the outer iterator
+          merge_occured = true;
+
+          // We need to start the list comparison over for the new it1 so break here
+          break;
         }
-      }
+      } // it2 loop
+
+      if (!merge_occured) // No merges so we need to manually increment the outer iterator
+        ++it1;
+
+    } // it1 loop
+  } // map loop
 
-      _partial_feature_sets[rank][map_num].clear();
+  /**
+   * All of the merges are complete and stored in a vector of lists. To make several
+   * of the sorting and tracking algorithms more straightforward, we will move these
+   * items into a vector of vectors instead.
+   */
+  _feature_count = 0;
+  for (auto map_num = decltype(_maps_size)(0); map_num < _maps_size; ++map_num)
```

Comment:
One of these days I'll figure out how to get rid of this whole "maps"
business so I can just use a range based for. Then I won't have this
problem. There are always going to be cases though were people need to
iterate over a range of integers and C++11 doesn't provide any clean ways
to just do that with an auto type...

On Wed, Jun 22, 2016 at 2:45 PM Derek Gaston <notifications@github.com>
wrote:

> In modules/phase_field/src/postprocessors/FeatureFloodCount.C
> <https://github.com/idaholab/moose/pull/7201#discussion_r68131634>:
>
> >
> > -      _partial_feature_sets[rank][map_num].clear();
> > +  /**
> > +   * All of the merges are complete and stored in a vector of lists. To make several
> > +   * of the sorting and tracking algorithms more straightforward, we will move these
> > +   * items into a vector of vectors instead.
> > +   */
> > +  _feature_count = 0;
> > +  for (auto map_num = decltype(_maps_size)(0); map_num < _maps_size; ++map_num)
>
> Yeah - I'm not the biggest fan of the syntax either. We (me @permcody
> <https://github.com/permcody> and @dschwen <https://github.com/dschwen> )
> already talked about this construction before. I personally prefer:
>
> decltype(foo) bar = 0
>
> That is valid syntax and matches up better with what we're used to. The
> problem is that sometimes foo is const... so this would actually make bar
> const as well and then you get a compile error on the bar++.
>
> auto bar = decltype(foo)(0)
>
> Doesn't have that issue.
>
> You could do:
>
> decltype(std::remove_const(foo)) bar = 0 or somesuch... but that's deeply
> unsatisfying as well ;-)
>
> auto bar = decltype(foo)(0) gets the job done and is simple (enough).
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/7201/files/1800330b4de20f90b511d453a72308208f05e5bb#r68131634>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AC5XIJjey9pMcPUHk0pcCshUiD9OixJ2ks5qOZ7YgaJpZM4I34-f>
> .
>